sympy/solvers/tests/test_solveset.py

```diff
@@ -1388,8 +1390,17 @@ def test_linear_eq_to_matrix():
     raises(ValueError, lambda: linear_eq_to_matrix(eqns3, [x, 1/a, y]))
     # 4) a nonlinear term is detected in the original expression
     raises(NonlinearError, lambda: linear_eq_to_matrix(Eq(1/x + x, 1/x), [x]))
+    # -- even when strict is False if it contains a generator
+    raises(NonlinearError, lambda: linear_eq_to_matrix(Eq(1/x + x, 1/x), [x],
+        strict=False))
+    # -- but not when the symbol-dependent term is not a generator
+    assert linear_eq_to_matrix(Eq(cos(x) + x, 1), [x], strict=False) == (
+        Matrix([[1]]), Matrix([[1 - cos(x)]]))
```

Comment:
I don't really understand this distinction about generators.

Reply:
All the decision making is in `_lin_eq2dict` at the end:

```python
    if not a.has_xfree(symset):  # no sym appears as arg in iterfreeargs
        return a, {}
```
After that, we know a sym appears but `x` might be appearing in `f(x)`, `1/x` or `x**2`. `x` only appears as a generator of the numerator in `x**2`, in the denominator of `1/x`, but not in `f(x)`--and I'm trying to use the Poly definition of "generator" as being an integer power of a given object that appears as a term or factor in a term of a flat expression. So the next line takes action if no symbol appears as a generator: if not being strict, then we treat the non-generated form `a` as a constant, else we raise the non-linear error:
```
    if not _has_gen(a, symset):
        # no generator in symset
        if not strict:
            # treat it as constant
            return a, {}
        raise PolyNonlinearError(filldedent('''
                nonlinearity can be ignored
                using `strict=False`'''))
```
If `a` is generated from one of the symbols then we have a nonlinear term
```
    # strict setting won't help in this case
    raise PolyNonlinearError('nonlinear in given symbols')
```

Reply:
How exactly is this "generator" handling different from just rejecting `Pow`?

Reply:
This version will recognize any power of a symbol as a generator in an expression's numerator or denominator.
```python
def _has_gen(self, s, p=None):
    """return True if any integer power of element in s appears as a
	term or factor in the numerator or denominator of self.

	Examples
	========

	>>> _has_gen(x**2, {x})
	True
	>>> _has_gen(1/x**2, {x})
	True
	>>> _has_gen(x**4, {x**2})
	True
	>>> _has_gen(x, {sqrt(x)})
	True
	>>> _has_gen(x**2, {x**3})
	False
	"""
    if self in s:
        return True
    if self.is_Add or self.is_Mul:
        return any(_has_gen(a, s, p) for a in self.args)
    b, c, e = _bce(self)
    if e == 1 and c != 1 and c.is_Integer and _has_gen(b, s, p):
        return True
    # check for powers
    if p is None:
        p = [_bce(i) for i in s if i.is_Pow or isinstance(i, exp)]
    for bi, ci, ei in p:
        if bi == b and ei == e and (c/ci).q == 1:
            return True
    return False
```